src/main/java/io/vertx/spi/cluster/consul/impl/cache/CacheMultiMap.java

```diff
@@ -76,7 +77,8 @@ public synchronized void remove(K key, V value) {
         ChoosableSet<V> choosableSet = cache.get(key);
         if (choosableSet == null) return;
         choosableSet.remove(value);
-        cache.put(key, choosableSet);
+        if (choosableSet.isEmpty()) cache.remove(key);
```

Comment:
` @Override
  protected <T> void sendOrPub(SendContextImpl<T> sendContext) {
    String address = sendContext.message.address();
    Handler<AsyncResult<ChoosableIterable<ClusterNodeInfo>>> resultHandler = asyncResult -> {
      if (asyncResult.succeeded()) {
        ChoosableIterable<ClusterNodeInfo> serverIDs = asyncResult.result();
        if (serverIDs != null && !serverIDs.isEmpty()) {
          sendToSubs(serverIDs, sendContext);
        } else {
          if (metrics != null) {
            metrics.messageSent(address, !sendContext.message.isSend(), true, false);
          }
          deliverMessageLocally(sendContext);
        }
      } else {
        log.error("Failed to send message", asyncResult.cause());
      }
    };`

fix is not really needed

Reply:
but it's also ok to have.